impedance/models/circuits/circuits.py

```diff
@@ -0,0 +1,446 @@
+            Lower and upper bounds on parameters. Defaults to bounds on all
+            parameters of 0 and np.inf, except the CPE alpha
+            which has an upper bound of 1
+
+        Returns
+        -------
+        self: returns an instance of self
+
+        """
+
+        # check that inputs are valid:
+        #    frequencies: array of numbers
+        #    impedance: array of complex numbers
+        #    impedance and frequency match in length
+
+        assert isinstance(frequencies, np.ndarray),\
+            'frequencies is not of type np.ndarray'
+        assert isinstance(frequencies[0], (float, int, np.int32, np.float64)),\
+            'frequencies does not contain a number'
+        assert isinstance(impedance, np.ndarray),\
+            'impedance is not of type np.ndarray'
+        assert isinstance(impedance[0], (complex, np.complex128)),\
+            'impedance does not contain complex numbers'
+        assert len(frequencies) == len(impedance),\
+            'mismatch in length of input frequencies and impedances'
+
+        if self.initial_guess is not None:
+            parameters, conf = circuit_fit(frequencies, impedance,
+                                           self.circuit, self.initial_guess,
+                                           self.constants, method=method,
+                                           bounds=bounds)
+            self.parameters_ = parameters
+            if conf is not None:
+                self.conf_ = conf
+        else:
+            # TODO auto calc guess
+            raise ValueError('no initial guess supplied')
+
+        return self
+
+    def _is_fit(self):
+        """ check if model has been fit (parameters_ is not None) """
+        if self.parameters_ is not None:
+            return True
+        else:
+            return False
+
+    def predict(self, frequencies, use_initial=False):
+        """ Predict impedance using a fit equivalent circuit model
+
+        Parameters
+        ----------
+        frequencies: numpy array
+            Frequencies
+
+        use_initial: boolean
+            If true and a fit was already completed,
+            use the initial parameters instead
+
+        Returns
+        -------
+        impedance: numpy array of dtype 'complex128'
+            Predicted impedance
+
+        """
+
+        # check that inputs are valid:
+        #    frequencies: array of numbers
+
+        assert isinstance(frequencies, np.ndarray),\
+            'frequencies is not of type np.ndarray'
+        assert isinstance(frequencies[0], (float, int, np.int32, np.float64)),\
+            'frequencies does not contain a number'
+
+        if self._is_fit() and not use_initial:
+            return eval(buildCircuit(self.circuit, frequencies,
+                                     *self.parameters_,
+                                     constants=self.constants, eval_string='',
+                                     index=0)[0])
```

Comment:
Instead of updating the globals() up top, I think we can just pass the circuit elements dictionary (which is a fantastic idea in terms of not importing a bunch of individual elements) directly to the eval function. So it'd just be `eval(....., circuit_elements)`. The downside of this is that we either have to add `s`, and `p` to the circuit elements dict or pass them in as well, but I think it might be better than updating the globals in the whole module in terms of readability/namespace?

Reply:
ooohh I had forgotten about the ability to pass a dictionary / namespace into `eval`, that would make it more readable for sure.  The upside of this implementation was that it was less modified code, but the readability has definitely suffered for it.  I'll test this out and see how it works.